crates/typst-utils/src/format.rs

```diff
@@ -0,0 +1,139 @@
+use std::{cmp::min, time::Duration};
+
+/// Returns number of `(days, hours, minutes, seconds, milliseconds, microseconds)`.
+fn get_duration_parts(duration: &Duration) -> (u16, u8, u8, u8, u16, u16) {
+    // In practice we probably don't need nanoseconds.
+    let micros = duration.as_micros();
+    let (millis, micros) = (micros / 1000, (micros % 1000) as u16);
+    let (sec, millis) = (millis / 1000, (millis % 1000) as u16);
+    let (mins, sec) = (sec / 60, (sec % 60) as u8);
+    let (hours, mins) = (mins / 60, (mins % 60) as u8);
+    let (days, hours) = ((hours / 24) as u16, (hours % 24) as u8);
+    (days, hours, mins, sec, millis, micros)
+}
+
+/// Format string using `days`, `hours`, `minutes`, `seconds`.
+fn format_dhms(days: u16, hours: u8, minutes: u8, seconds: u8) -> String {
+    match (days, hours, minutes, seconds) {
+        (0, 0, 0, s) => format!("{s:2} s"),
+        (0, 0, m, s) => format!("{m:2} m {s:2} s"),
+        (0, h, m, s) => format!("{h:2} h {m:2} m {s:2} s"),
+        (d, h, m, s) => format!("{d:3} d {h:2} h {m:2} m {s:2} s"),
+    }
+}
+
+/// Format string starting with number of days and going bigger from there.
+fn time_starting_with_seconds(duration: &Duration) -> String {
+    let (days, hours, minutes, seconds, _, _) = get_duration_parts(duration);
+    format_dhms(days, hours, minutes, seconds)
+}
+
+/// Format string starting with number of milliseconds and going bigger from there.
+/// `precision` is how many digits of microseconds from left will be preserved.
+/// Note that this function will always remove all trailing zeros for microseconds.
+fn time_starting_with_ms_with_precision(duration: &Duration, precision: usize) -> String {
+    let digits = precision;
+    let (d, h, m, s, ms, mcs) = get_duration_parts(duration);
+    let mcs_digits = {
+        let mut mcs = mcs;
+        println!("before {mcs}");
+        while mcs % 10 == 0 && mcs != 0 {
+            mcs /= 10
+        }
+        println!("after {mcs}");
+        mcs.to_string().len()
+    };
+    let ms = format!(
+        "{1:.0$} ms",
+        if mcs != 0 { min(digits, mcs_digits) } else { 0 },
+        ms as f32 + mcs as f32 / 1000.0
+    );
+    match (d, h, m, s) {
+        (0, 0, 0, 0) => ms.to_string(),
+        (d, h, m, s) => format_dhms(d, h, m, s),
+    }
+}
```

Comment:
Something like this seems better - see [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=afa07ceb3c758d96956cb9e8724c744a):

(note: there's more text below the suggestion)

```suggestion
/// Format string starting with number of milliseconds and going bigger from there.
/// `precision` is how many digits of microseconds from left will be preserved.
/// Note that this function will always remove all trailing zeros for microseconds.
fn time_starting_with_ms_with_precision(duration: &Duration, precision: u16) -> String {
    let (d, h, m, s, ms, mcs) = get_duration_parts(duration);
    match (d, h, m, s) {
        (0, 0, 0, 0) => {
            let ms = ms as f64 + mcs as f64 / 1000.0;
            let ms = (ms * 10.0_f64.powi(precision as i32)).round() / 10.0_f64.powi(precision as i32);
            format!("{ms} ms")
        }
        (d, h, m, s) => format_dhms(d, h, m, s),
    }
}
```

Although it may be a good idea to extract the 10^precision shenanigans into a separate `typst-utils` function, as this is also used for `repr`:

https://github.com/typst/typst/blob/d84a7e423b53e9b5740720270b132da60c8a3fea/crates/typst/src/foundations/repr.rs#L86-L89

Reply:
Ah, yes, the "measure, cut+round, reset" method. I just don't like multiplying by non-small numbers, worried about going over the limit. But the algorithm is very simple and gets the job done. This is at least a concern in the `repr.rs`.

Reply:
In principle there isn't much risk of overflow cuz float (at worst you get some infinity somewhere), though there are some potential mitigations that could be done if needed. However, we only use this with p = 2 right now so should be ok.

Reply:
Oh, heh, it already triggered it [here](https://github.com/Andrew15-5/typst/actions/runs/10405315865/job/28815783456). I run all tests locally and didn't have any failed ones. That's kinda strange...

Reply:
It was because `value` is already infinity. I think this can be an edge case. Now tests pass.